### PR TITLE
check structure when merging blocks

### LINKED ISSUE #6564: check structure when merging blocks
## Enhancement
When merging blocks, we need to check their structure is the same.

### PR DESCRIPTION
**Pull request by mengxin9014**
### What problem does this PR solve?

Issue Number: close #6564

Problem Summary:

### What is changed and how it works?
add structure check when merging blocks.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Core/Block.cpp [hunk 1]
```diff
@@ -543,8 +543,12 @@ Block mergeBlocks(Blocks && blocks)
 
     for (size_t i = 1; i < blocks.size(); ++i)
     {
-        if (likely(blocks[i].rows()) > 0)
+        if (likely(blocks[i].rows() > 0))
         {
+            if (unlikely(!blocksHaveEqualStructure(first_block, blocks[i])))
+            {
+                throw Exception("Cannot merge different structure blocks.", ErrorCodes::LOGICAL_ERROR);
+            }
```
Comment (SeaRise) [inline dbms/src/Core/Block.cpp:551]: Maybe we just need to check under debug mode?
```suggestion
            assert(blocksHaveEqualStructure(first_block, blocks[i]));
```

Comment (SeaRise) [reply dbms/src/Core/Block.cpp:551]: Or use RUNTIME_CHECK
```suggestion
            RUNTIME_CHECK(blocksHaveEqualStructure(first_block, blocks[i]), "Cannot merge different structure blocks.");
```

**Review by yibin87 — APPROVED** [review]
LGTM

File: dbms/src/Core/Block.cpp (hunk 1 above)
Comment (mengxin9014) [reply dbms/src/Core/Block.cpp:551]: > Or use RUNTIME_CHECK

OK.

File: dbms/src/Core/Block.cpp [hunk 2]
```diff
@@ -543,8 +543,12 @@ Block mergeBlocks(Blocks && blocks)
 
     for (size_t i = 1; i < blocks.size(); ++i)
     {
-        if (likely(blocks[i].rows()) > 0)
+        if (likely(blocks[i].rows() > 0))
         {
+            if (unlikely(!blocksHaveEqualStructure(first_block, blocks[i])))
```
Comment (bestwoody) [inline dbms/src/Core/Block.cpp:548]: why the same block case will be encountered? Are there some expected cases that it needs merge identical blocks?

Comment (xzhangxian1008) [inline dbms/src/Core/Block.cpp]: maybe we can use `RUNTIME_CHECK_MSG`

Comment (mengxin9014) [reply dbms/src/Core/Block.cpp]: > why the same block case will be encountered? Are there some expected cases that it needs merge identical blocks?

This is for protection, anyone can call mergeBlocks, if there is a different structure in the input blocks, the result will be problematic.  Its function is  similar here https://github.com/pingcap/tiflash/blob/fececd9375bd3b5de2318788bb08fa84e7c5deed/dbms/src/Interpreters/Aggregator.cpp#L2042

File: dbms/src/Core/Block.cpp [hunk 3]
```diff
@@ -543,8 +543,9 @@ Block mergeBlocks(Blocks && blocks)
 
     for (size_t i = 1; i < blocks.size(); ++i)
     {
-        if (likely(blocks[i].rows()) > 0)
+        if (likely(blocks[i].rows() > 0))
         {
+            RUNTIME_CHECK_MSG(blocksHaveEqualStructure(first_block, blocks[i]), "Cannot merge different structure blocks.");
```
Comment (xzhangxian1008) [inline dbms/src/Core/Block.cpp:548]: Why we need to check structure? What will happen when two blocks' structure are different?

File: dbms/src/Core/Block.cpp (hunk 1 above)
Comment (mengxin9014) [reply dbms/src/Core/Block.cpp]: > Or use RUNTIME_CHECK

done.

File: dbms/src/Core/Block.cpp (hunk 2 above)
Comment (mengxin9014) [reply dbms/src/Core/Block.cpp]: > why the same block case will be encountered? Are there some expected cases that it needs merge identical blocks?

merging identical blocks you can refer https://github.com/pingcap/tiflash/issues/6533

File: dbms/src/Core/Block.cpp (hunk 3 above)
Comment (mengxin9014) [reply dbms/src/Core/Block.cpp:548]: > Why we need to check structure? What will happen when two blocks' structure are different?

This is for protection, anyone can call mergeBlocks, if there is a different structure in the input blocks, the result will be problematic.  Its function is  similar here https://github.com/pingcap/tiflash/blob/fececd9375bd3b5de2318788bb08fa84e7c5deed/dbms/src/Interpreters/Aggregator.cpp#L2042

Comment (mengxin9014) [reply dbms/src/Core/Block.cpp:548]: > Why we need to check structure? What will happen when two blocks' structure are different?

if the column of block[0] is nullable, block[1] is not nullable.  the merging will transfer the column of block[1] to nullable and insert to the column block[0]. You can't be sure if the column we need will be nullable or not.

File: dbms/src/Core/Block.cpp (hunk 1 above)
Comment (mengxin9014) [reply dbms/src/Core/Block.cpp]: > Maybe we just need to check under debug mode?

done.

**Review by SeaRise — APPROVED** [review]
lgtm

**Comment by xzhangxian1008** [comment]
lgtm

**Review by bestwoody — APPROVED** [review]
lgtm

**Pull request merged**